libreoffice: add update script and 7.2.5 -> 7.3.3

tricktron opened the pull request:
###### Description of changes
Fixes https://github.com/NixOS/nixpkgs/issues/174063.

The `7.2.5` version is already outdated and 404'ing. This pr adds the following:
- Update script with [nixt](https://github.com/nix-community/nixt/) tests to ease the update process: just run `nix-shell maintainers/scripts/update.nix --argstr package libreoffice` to update to the latest stable version from https://download.documentfoundation.org/libreoffice/stable/. Furthermore, this may even enable and trigger simple https://github.com/ryantm/nixpkgs-update package pr to proactively update to the latest stable version.
- Updates from 7.2.5 -> 7.3.3
- Minor general darwin improvements

@NixOS/darwin-maintainers: I tested this on aarch64-darwin. Could someone give this a test ride on x86_64-darwin?

One question remains:
- What is needed so that the darwin platforms and versions also show up on https://search.nixos.org/packages? How do you handle different versions for darwin vs. linux on https://search.nixos.org/packages? Do I have to give my derivation another `pname`?

Edit: Answered here https://github.com/NixOS/nixos-search/issues/485.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tricktron committed:
libreoffice: add update-script for darwin

tricktron committed:
libreoffice: refactor and document version reset workaround

tricktron committed:
libreoffice: general darwin improvements

- throw error if platform is not supported
- fix wrong pre/postInstallPhase hook -> pre/postInstall
- switch to stdenvNoCC because we don't build anything
- run all phases by default
- wrap binary using `open -na $out/Applications/${appName} --args "$@"`

tricktron committed:
libreoffice: 7.2.5 -> 7.3.3

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -27754,7 +27754,7 @@ with pkgs;
 
   libreoffice = if stdenv.isDarwin
-    then callPackage ../applications/office/libreoffice/darwin.nix {}
+    then callPackage ../applications/office/libreoffice/darwin/darwin.nix {}
```
```suggestion
    then callPackage ../applications/office/libreoffice/darwin { }
```

tricktron replied:
@SuperSandro2000 This only works if rename `darwin.nix` -> `default.nix`. Did you mean that?

SuperSandro2000 replied:
yeah, I implied that

tricktron replied:
done

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/update-utils.nix:
```diff
@@ -0,0 +1,54 @@
+    (builtins.replaceStrings [ oldVersion ] [ newVersion ] dmgUrl) + ".sha256";
+
+  # nullHash :: String
+  nullHash = "0000000000000000000000000000000000000000000000000000000000000000";
```
Use lib.fakeSha256

tricktron replied:
Done, thanks.

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/update-utils.nix:
```diff
@@ -0,0 +1,54 @@
+{ lib }:
+let
+  # extractLatestVersionFromHtml :: String -> String
+  extractLatestVersionFromHtml = htmlString:
```
Why are we not using something easy to use like pup?

tricktron replied:
I wanted to write the complete update-script in nix because it integrates much better with changes to the derivation. In addition, I wanted to try out a test-driven approach with nixt and I quite like it.

 If I were to write it in bash, pup would obviously be a good choice.

SuperSandro2000 replied:
update scripts are usually not written in bash for the reasons I already mentioned.

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/update-utils.nix:
```diff
@@ -0,0 +1,54 @@
+  # extractSha256FromHtml :: String -> String
+  extractSha256FromHtml = htmlString:
+    let
+      sha256 = (builtins.match ".*([0-9a-fA-F]{64}).*" htmlString);
```
This could be very easily done with grep and be easy to debug.

tricktron replied:
Same reasons as above of trying out a nix only approach.

In my opinion this is also easy to debug because there are tests and the nix-repl is also quite helpful to try out/manually test these functions.

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/update-test.nix:
```diff
@@ -0,0 +1,47 @@
+# run the tests with nixt <absolutePath to parent dir> -v
```
I don't think we need this at all. We can simply locally revert a last update and see if the update script works. This also gives us a very very false sense of security because there website could easily change tomorrow and we would have green testing.

tricktron replied:
I added an integration test for the website to solve the false sense of security issue.

SuperSandro2000 replied:
Usually there are no updater tests because they can be out of date with every update and potentially break especially when parsing html. Best to subscribe to another update channel like RSS or github releases and make sure through that that nothing is missed.

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/darwin.nix:
```diff
@@ -0,0 +1,84 @@
+      ''
+        #!/usr/bin/env nix-shell
+        #!nix-shell -i bash -p common-updater-scripts
+        set -o errexit
+        set -o nounset
+        set -o pipefail
```
```suggestion
        set -eou pipefail
```
short form

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/darwin.nix:
```diff
@@ -0,0 +1,84 @@
+        set -o pipefail
+
+        # reset version first so that both platforms are always updated and in sync
+        update-source-version libreoffice 0 ${nullHash} --system=aarch64-darwin
+        update-source-version libreoffice ${newVersion} ${newAarch64Sha256} --system=aarch64-darwin
+        update-source-version libreoffice 0 ${nullHash} --system=x86_64-darwin
+        update-source-version libreoffice ${newVersion} ${newX86_64Sha256} --system=x86_64-darwin
```
We shouldn't need to reset the version first. This can be done in one go.

SuperSandro2000 replied:
Then your update script is not great. Usually the version is updated and then the hash in the same run.

SuperSandro2000 commented on pkgs/applications/office/libreoffice/darwin/darwin.nix:
```diff
@@ -0,0 +1,84 @@
+}:
+
+let
+  appName = "LibreOffice.app";
+  scriptName = "soffice";
+  version = "7.3.3";
+  baseUrl = "https://download.documentfoundation.org/libreoffice/stable";
+  dist = {
+    aarch64-darwin = rec {
+      arch = "aarch64";
+      archSuffix = arch;
+      url = "${baseUrl}/${version}/mac/${arch}/LibreOffice_${version}_MacOS_${archSuffix}.dmg";
+      sha256 = "50ed3deb8d9c987516e2687ebb865bca15486c69da79f1b6d74381e43f2ec863";
+    };
+
+    x86_64-darwin = rec {
+      arch = "x86_64";
+      archSuffix = "x86-64";
+      url = "${baseUrl}/${version}/mac/${arch}/LibreOffice_${version}_MacOS_${archSuffix}.dmg";
+      sha256 = "fb2f9bb90eee34a22af3a2bf2854ef5b76098302b3c41d13d4f543f0d72b994f";
+    };
```
```suggestion
  appName = "LibreOffice.app";
  scriptName = "soffice";
  version = "7.3.3";
  dist = {
    aarch64-darwin = rec {
      arch = "aarch64";
      archSuffix = arch;
      url = "https://download.documentfoundation.org/libreoffice/stable/${version}/mac/${arch}/LibreOffice_${version}_MacOS_${archSuffix}.dmg";
      sha256 = "50ed3deb8d9c987516e2687ebb865bca15486c69da79f1b6d74381e43f2ec863";
    };

    x86_64-darwin = rec {
      arch = "x86_64";
      archSuffix = "x86-64";
      url = "https://download.documentfoundation.org/libreoffice/stable/${version}/mac/${arch}/LibreOffice_${version}_MacOS_${archSuffix}.dmg";
      sha256 = "fb2f9bb90eee34a22af3a2bf2854ef5b76098302b3c41d13d4f543f0d72b994f";
    };
```

I don't even know from where arch and archSuffix are coming. This is not easily debuggable when the URL changes.

tricktron committed:
libreoffice: apply suggestions from code review

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

tricktron committed:
libreoffice: add integration test and use `lib.fakeSha256`

uri-canva commented:
Tested on `x86_64-darwin`, both the update script and libreoffice seem to work fine.

tricktron commented:
@ofborg build libreoffice

tricktron commented:
Follow-up to the discussion here https://github.com/NixOS/nixos-search/issues/485:

Is there any downside if we rename this to libreoffice-darwin in all-packages.nix because this derivation is completely different from the linux one?

The advantage would be a clear separation and visibility on https://search.nixos.org/packages.

uri-canva commented:
They linked an example of a pattern that works in that issue: https://github.com/NixOS/nixpkgs/blob/5ce6597eca7d7b518c03ecda57d45f9404b5e060/pkgs/applications/networking/instant-messengers/teams/default.nix#L27-L34=.

tricktron committed:
libreoffice: move darwin to separate libreoffice-darwin package

The linux and darwin derivations are completely different. Therefore, it
makes more sense to treat them as two separate, independent derivations.

tricktron commented:
@ofborg build libreoffice-darwin

tricktron committed:
Revert "libreoffice: move darwin to separate libreoffice-darwin package"

Introducing a new name for each complex darwin derivation does not
scale.

This reverts commit b207ef980751b2dfe2f222fbd7dbc854f9bd90bf.

tricktron committed:
libreoffice: add darwin to meta platforms and extract to common.nix

As a result, the darwin derivations will show up on
https://search.nixos.org/packages.

tricktron committed:
libreoffice: pass explicit file to update-source-version function and
fix import typo

tricktron commented:
> They linked an example of a pattern that works in that issue: https://github.com/NixOS/nixpkgs/blob/5ce6597eca7d7b518c03ecda57d45f9404b5e060/pkgs/applications/networking/instant-messengers/teams/default.nix#L27-L34=.

I know because I wrote that code😄.

tricktron commented:
@ofborg build libreoffice

tricktron commented:
@7c6f434c Just for your information: I extracted the meta attribute to `common.nix` and added darwin to supported platforms so that it will show up on https://search.nixos.org/packages.

tricktron commented:
@SuperSandro2000 I always appreciated your detailed feedback and experience on prs.

However, I think that the discussion around *update script written in nix with tests* vs. *update script written in bash* is not going into a fruitful direction.

Feedback like that:
> Then your update script is not great. Usually the version is updated and then the hash in the same run.

is not constructive and leads nowhere (especially when the update script is currently working as intended).

I propose to move forward with the current working update script written in nix. Time will tell if it is indeed as fragile as you write it to be. If it is then the burden is on me since I am the maintainer and I'll gladly fix it and convert it to a bash script using pup.

7c6f434c commented:
1. I do not think splitting is bad because yours is `libreoffice-bin` not really `libreoffice` by the common attribute naming nomenclature.

2. The very idea of imperative code in Nix inside Nixpkgs is not what people expect. This in itself is a valid criticism.

tricktron commented:
@7c6f434c I agree that `libreoffice-bin` is more accurate according to the nomenclature but it is not commonly used for darwin. Therefore I would keep it at libreoffice to keep it simple. Are you ok with sharing the meta attribute and keeping it at libreoffice or do you want me to introduce `libreoffice-bin`?

SuperSandro2000 commented:
> 1. I do not think splitting is bad because yours is `libreoffice-bin` not really `libreoffice` by the common attribute naming nomenclature.

If there is a source build version and a downloaded and patched version, then both names should exist. If there is only one version the name is libreoffice and the information that it is not source build should eventually be in meta when people figured out how they want to do that.

> I propose to move forward with the current working update script written in nix.

I don't agree with that. There are good reasons why update scripts are not written in nix and have no tests. They are a bit flimsy and especially the ones that parse websites are fragile and cannot be made future proof. Ideally we would want to use some kind of API instead.

> is not constructive and leads nowhere (especially when the update script is currently working as intended).

You described a bug in the update script to me and I wanted to imply that it should be fixed because most other update scripts also can do that in one run.

PS: Just the fact that something works does not make it a good solution to a problem. It can work and at the same time be an ugly solution which shouldn't really be done.

7c6f434c commented:
I don't really like an outcome where I am listed among maintainers of something that I a) have never been interested in, b) find too magic (and I like tricks, it's just that this goes too far for Nixpkgs in my opinion)

tricktron commented:
> I don't really like an outcome where I am listed among maintainers of something that I a) have never been interested in, b) find too magic (and I like tricks, it's just that this goes too far for Nixpkgs in my opinion)

What magic are you talking about?

tricktron committed:
libreoffice: rename libreoffice -> libreoffice-bin

tricktron commented:
> If there is a source build version and a downloaded and patched version, then both names should exist. If there is only one version the name should be libreoffice and the information that it is not source build should eventually be in meta when people figured out how they want to do that.

Done I renamed the darwin version to `libreoffice-bin`.

> I don't agree with that. There are good reasons why update scripts are not written in nix and have no tests. They are a bit flimsy and especially the ones that parse websites are fragile and cannot be made future proof. Ideally we would want to use some kind of API instead.

I completely agree that we should always use an api if available.
However, I am not convinced by your argument: *We have always written an update-script in bash, therefore this is the only right way to do that* and *if there is no api, you should not write an update script because parsing html is fragile and not future proof*. What I am doing here is merely the following:

1. I precompute the latest version and the sha256 by parsing some html (because I have not found an api) using nix.
2. I use this precomputed values in the update bash script using the recommended `update-source-version` script to update the version.
3. Since I update two platforms (aarch64-darwin and x86_64-darwin) I need to use a workaround by resetting the version first, which is also used in the [sublime4](https://github.com/NixOS/nixpkgs/blob/634141959076a8ab69ca2cca0f266852256d79ee/pkgs/applications/editors/sublime/4/common.nix#L133-L139) derivation.

So I really don't understand why step number 1 is causing so much problems. Besides, the update-script is mostly used by maintainers (me) and if there are automated tests to help me spot changes to my source early, I really like that.

> You described a bug in the update script to me and I wanted to imply that it should be fixed because most other update scripts also can do that in one run.

No I did not describe a bug. You just commented that *We shouldn't need to reset the version first. This can be done in one go.* and I then pointed out that I couldn't make it work without the abovementioned workaround.

> PS: Just the fact that something works does not make it a good solution to a problem. It can work and at the same time be an ugly solution which shouldn't really be done.
I completely agree. But in my opinion it is always better to: first make it work and then beautiful. As I already said, if the update-script breaks in the future (because it is ugly, which you imply here), then I fix it and make it better.

I would love to get some other honest opinions on that topic. I am actively defending my rights as a maintainer to decide about the implementation details of the update script. Am I being too stubborn? @NixOS/nixpkgs-committers.

tricktron commented:
@ofborg build libreoffice-bin

7c6f434c commented:
> What magic are you talking about?

Nix code with intentionally non-reproducible evaluation. Moreover, `updateScript` that determines upstream version while being evaluated, not while being run.

piegamesde commented:
@tricktron You just pinged over 150 people. Please don't do that. Since GitHub does not allow bulk-unsubscribing people from issue and in order to keep spamming them at a minimum, I am going to close this PR. Please open a new one.

tricktron commented:
@piegamesde ups, sorry pinged the wrong group:sweat_smile:. Will do.

tricktron commented:
New pr is here https://github.com/NixOS/nixpkgs/pull/175160

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/updatescript-written-in-nix-considered-bad-practice-vs-bash/19379/1
